### PR TITLE
[PPP-4431] - IBM ASoC DAST Scan: Improper Restriction of Excessive Au…

### PR DESCRIPTION
**Pull request by bcostahitachivantara**
…thentication Attempts (/pentaho/j_spring_security_check: j_password)

@pentaho-lmartins 
@pentaho/tatooine

### CODE REVIEW

**Comment by buildguy** [comment]
### :warning: Build finished in 41m 49s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,core,extensions
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 1 test(s) failed:

<details><summary><b>org.pentaho.platform.web.http.security.PreventBruteForceUsernamePasswordAuthenticationFilterTest.testAttemptAuthenticationWithoutBlock</b> (click to expand)</summary><p>

```
${result.errorDetails}
```

</p></details>

  
Tests run: **2583**, Failures: **1**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/platform/job/PR-4669/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 34m 30s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,core,extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **2583**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/platform/job/PR-4669/2/testReport)

 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 32m 52s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,core,extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **2583**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/platform/job/PR-4669/3/testReport)

 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 41m 43s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,core,extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **2583**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/platform/job/PR-4669/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security.xml [hunk 1]
```diff
@@ -253,19 +253,32 @@
...
+  </bean>
+
+  <bean id="preventBruteForceAuthenticationFailureListener"
+        class="org.pentaho.platform.web.http.security.PreventBruteForceAuthenticationFailureListener" >
+    <constructor-arg ref="loginAttemptService"/>
+  </bean>
+
   <bean id="authenticationProcessingFilter"
-        class="org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter">
+        class="org.pentaho.platform.web.http.security.PreventBruteForceUsernamePasswordAuthenticationFilter">
```
Comment (befc) [inline assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security.xml]: Not sure it's good practice to have "PreventBruteForce" in the name of the classes, usually security sensitive information shouldn't stay so exposed as to be even mentioned in a configuration file.

Comment (bcostahitachivantara) [reply assemblies/pentaho-solutions/src/main/resources/pentaho-solutions/system/applicationContext-spring-security.xml]: I understand your concern but the name of class should represent what class does. 

Beside that this configuration is on server side and is not exposed to the outside. This is open source code so even with a different name everyone could read the content and inffer what class does.

**Comment by bcostahitachivantara** [comment]
have saw that PR add some code smell i am going do a new commit to fix that

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 1h 6m 28s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,core,extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **2583**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/platform/job/PR-4669/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 42m 51s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,core,extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **2583**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/platform/job/PR-4669/6/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 42m 19s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,core,extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **2583**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/platform/job/PR-4669/7/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by pedrofvteixeira** [comment]
Overall the PR looks good. 

I would double-check / ensure we are good in applying changes to a spring xml file known to be editable by customers per their needs ( i.e. `applicationContext-spring-security.xml` )

List of checks I would consider ( also pinging @pentaho-lmartins here ) : 

 - Are we ok in doing any changes to `applicationContext-spring-security.xml`, knowing upfront that customer will hold their own edited version of this file ?
 - Is this bug meant to be backported ? If so, what are the changes we would be backporting, when it comes to `applicationContext-spring-security.xml` ?
 - What happens should a customer upgrade to the newer version ( or get the SP with this fix ), but keeps their "old" `applicationContext-spring-security.xml` ? 
    - Will anything break, or will it simply not have this feature applied ?
- What is the expectation in regards to the application of this fix, in scenarios where customer has enabled federated authentication, knowing that both CAS and SAML carry their own spring xml files that override parts of `applicationContext-spring-security.xml` ?
  - CAS: https://help.pentaho.com/Documentation/8.1/Setup/Administration/User_Security/Implement_Advanced_Security/060/000
  - SAML: https://help.pentaho.com/Documentation/9.0/Setup/SAML_security

**Comment by befc** [comment]
The implementation of the feature seems good.
Besides the questions raised by @pedrofvteixeira, I have one more.

- Does the UI message change when the attempts are expired? If it doesn't, let's imagine, in a scenario where we have a "real" user, that just forgot his password. He may be trying to remember it and trying to login at the same time. If he has no information in the UI that his attempts have expired, he may just keep trying to infinity, maybe even finally input the correct password at the stage his attempts have expired, and he will just think his correct password is not the correct one. Due to this scenario, I would advice to change the message prompted to the user. Seems also to be a common message to present on the web in such scenarios.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 1h 6m 38s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,core,extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **2583**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/platform/job/PR-4669/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 43m 2s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
api,assemblies/pentaho-solutions,assemblies/pentaho-war,core,extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **2581**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/platform/job/PR-4669/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Pull request merged**